drivers/SmartThings/philips-hue/src/handlers/attribute_emitters.lua

```diff
@@ -125,30 +112,120 @@ local function _emit_light_events_inner(light_device, light_repr)
   end
 end
 
+function AttributeEmitters.connectivity_update(child_device, zigbee_status)
+  if zigbee_status.status == "connected" then
+    child_device.log.info_with({hub_logs=true}, "Light status event, marking device online")
+    child_device:online()
+    child_device:set_field(Fields.IS_ONLINE, true)
+  elseif zigbee_status.status == "connectivity_issue" then
+    child_device.log.info_with({hub_logs=true}, "Light status event, marking device offline")
+    child_device:set_field(Fields.IS_ONLINE, false)
+    child_device:offline()
+  end
+end
+
+function AttributeEmitters.emit_contact_sensor_attribute_events(sensor_device, sensor_info)
+  if sensor_device == nil or (sensor_device and sensor_device.id == nil) then
+    log.warn("Tried to emit attribute events for a device that has been deleted")
+    return
+  end
+
+  if sensor_info.power_state then
+    log.debug(true, "emit power")
+    sensor_device:emit_event(capabilities.battery.battery(st_utils.clamp_value(sensor_info.power_state.battery_level, 0, 100)))
+  end
+
+  if sensor_info.tamper_reports then
```

Comment:
Is opening the battery enclosure what the device considers tampering with it?

Reply:
Yup.